crates/balance-alerts-layer/src/lib.rs

```diff
@@ -0,0 +1,172 @@
+use alloy::transports::{Transport, TransportResult};
+
+/// Configuration for the BalanceAlertLayer
+#[derive(Debug, Clone, Default)]
+pub struct BalanceAlertConfig {
+    /// Address to periodically check the balance of
+    pub watch_address: Address,
+    /// Threshold at which to log a warning
+    pub warn_threshold: Option<U256>,
+    /// Threshold at which to log an error
+    pub error_threshold: Option<U256>,
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct BalanceAlertLayer {
+    config: BalanceAlertConfig,
+}
+
+/// A ProviderLayer that can be added to an alloy Provider
+/// to log warnings and errors when the balance of a given address
+/// falls below certain thresholds.
+///
+/// This checks the balance after every transaction sent via send_transaction
+/// and errors, warns or trace logs accordingly
+///
+/// # Examples
+/// ```ignore
+/// let provider = ProviderBuilder::new()
+///     .layer(BalanceAlertLayer::new(BalanceAlertConfig {
+///         watch_address: wallet.default_signer().address(),
+///         warn_threshold: parse_ether("0.1")?,
+///         error_threshold: parse_ether("0.1")?,
+///     }));
+/// ```
+impl BalanceAlertLayer {
+    pub fn new(config: BalanceAlertConfig) -> Self {
+        Self { config }
+    }
+}
+
+impl<P, T> ProviderLayer<P, T, Ethereum> for BalanceAlertLayer
+where
+    P: Provider<T>,
+    T: Transport + Clone,
+{
+    type Provider = BalanceAlertProvider<P, T>;
+
+    fn layer(&self, inner: P) -> Self::Provider {
+        BalanceAlertProvider::new(inner, self.config.clone())
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct BalanceAlertProvider<P, T> {
+    inner: P,
+    config: BalanceAlertConfig,
+    _pd: PhantomData<fn() -> T>,
+}
+
+impl<P, T> BalanceAlertProvider<P, T>
+where
+    P: Provider<T>,
+    T: Transport + Clone,
+{
+    #[allow(clippy::missing_const_for_fn)]
+    fn new(inner: P, config: BalanceAlertConfig) -> Self {
+        Self { inner, config, _pd: PhantomData }
+    }
+}
+
+#[async_trait::async_trait]
+impl<P, T> Provider<T> for BalanceAlertProvider<P, T>
+where
+    P: Provider<T>,
+    T: Transport + Clone,
+{
+    #[inline(always)]
+    fn root(&self) -> &RootProvider<T> {
+        self.inner.root()
+    }
+
+    /// Broadcasts a raw transaction RLP bytes to the network.
+    ///
+    /// This override checks the watched address after sending the transaction and
+    /// logs a warning or error if the balance falls below the configured thresholds.
+    ///
+    /// See [`send_transaction`](Self::send_transaction) for more details.
+    async fn send_raw_transaction(
+        &self,
+        encoded_tx: &[u8],
+    ) -> TransportResult<PendingTransactionBuilder<T, Ethereum>> {
+        let res = self.inner.send_raw_transaction(encoded_tx).await;
+        let balance = self.inner.get_balance(self.config.watch_address).await?;
+
+        if balance < self.config.error_threshold.unwrap_or(U256::MAX) {
```

Comment:
Shouldn't this be `unwrap_or(0)`? If config is not set we wouldn't want to log errors/warnings

Would be good to have a test for the no config case also

Reply:
Yep that does makes sense rather than for always to alert when there is no config. 

Will update and add a test

Reply:
@willpote Updated to default to no warnings/errors and adds a test showing this